### PR TITLE
Initial refactor

### PR DESCRIPTION
**Pull request by blopax**
Initial refactor using https://github.com/EParisot/Patate as a starting point.
Refactor this repo aimed to:

    create clearer architecture using folders and modular classes
    have a baseline to build pipelines
    add useful scripts
    make it pep8 compliant

### CODE REVIEW

File: get_data/get_data_xbox_pad.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import Adafruit_PCA9685
+import xbox
+# noinspection PyUnresolvedReferences
+from picamera import PiCamera
+# noinspection PyUnresolvedReferences
+from picamera.array import PiRGBArray
+# noinspection PyUnresolvedReferences
+from PIL import Image
+
+from utils.const import *
```
Comment (Talasta) [inline get_data/get_data_xbox_pad.py]: A eviter les import *

File: get_data/get_data_xbox_pad.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        elif x_cursor <= 1:
+            self.direction = DIRECTION_R_M
+        self.label[1] = round(x_cursor, 2)
+
+        # Set motor direction and speed
+        self.pwm.set_pwm(0, 0, self.direction)
+        self.pwm.set_pwm(1, 0, self.speed)
+
+
+if __name__ == "__main__":
```
Comment (Talasta) [inline get_data/get_data_xbox_pad.py]: Idealement il faudrait ne pas avoir de declaration de class dans les fichiers de scripts, c'est plus lisible.

File: process_data/transform_data.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+import os
+import argparse
```
Comment (Talasta) [inline process_data/transform_data.py]: Les import par categories (natif, installé, local) puis ordre alphabetique. ;)

File: race.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import numpy as np
+import time
+
+import argparse
+import Adafruit_PCA9685
+# noinspection PyUnresolvedReferences
+from keras.models import load_model
+
+from pivideostream import PiVideoStream
+from utils.const import *
```
Comment (Talasta) [inline race.py]: Pareil, `import *` c'est dangereux

File: train_data/model_params_setter.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#  Patate/Data_processing/Training/3_directions_models_Opti_speed.ipynb
+
+import keras.backend as K
+from keras.layers import *
```
Comment (Talasta) [inline train_data/model_params_setter.py]: Encore un :)

File: train_data/train.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    def train(self):
+        self.model = Model(inputs=self.model_inputs, outputs=self.model_outputs)
+        self.model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
+        print(self.model.summary)
+        best_checkpoint = keras.callbacks.ModelCheckpoint(self.model_name, monitor='val_loss', verbose=1,
+                                                          save_best_only=True, mode='min')
+        self.model.fit(self.images, [self.labels_speed, self.labels_directions],
+                       batch_size=64, epochs=100, validation_split=0.2,
+                       verbose=1, callbacks=[best_checkpoint])
```
Comment (Talasta) [inline train_data/train.py]: Utilisez tensorboard :)

File: train_data/validation.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Validation_3.ipynb
+
+# TODO: use sklearn to get metrics between predictions and truth (self.labels_directions, self.label_speed)
```
Comment (Talasta) [inline train_data/validation.py]: Decider sur quel metrics vous voulez vous baser (accuracy, precision, recall, f1 score) ?

File: utils/check_pycodestyle.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+input='.'
+if [ $1 ]; then
+    input=$1
+fi
+
+pycodestyle --max-line-length=120 --exclude="**/*.h5" ${input}
```
Comment (Talasta) [inline utils/check_pycodestyle.sh]: C'est 80 normalement la norme sur python.

File: utils/hardware_tests.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+"""
+Aim of this file: concentrate all tests that check if hardware working fine:
+motors, direction, head, camera, other captors...
+These tests can potentially be linked to some calibration on hardware set.
+In this case it should be clear.
+"""
+
```
Comment (Talasta) [inline utils/hardware_tests.py]: C'est un genre d'unitesting ?

File: train_data/train.py (hunk 6 above)
Comment (blopax) [reply train_data/train.py]: Bien cu je le rajoute dans la todo

File: train_data/validation.py (hunk 7 above)
Comment (blopax) [reply train_data/validation.py]: yes il y a tout un topic dessus et comment utiliser pour automatiser la selection. on peut aussi en avoir plusieurs et afficher le tout sur le web (ou a defaut un doc reference).

File: utils/check_pycodestyle.sh (hunk 8 above)
Comment (blopax) [reply utils/check_pycodestyle.sh]: haha l'oeil de lynx :) c vrai. si dans la team les gens preferent 80 je m'adapterai. J'avoue que j aime bien 120 comme format mais c'est juste une preference personnelle.

File: utils/hardware_tests.py (hunk 9 above)
Comment (blopax) [reply utils/hardware_tests.py]: Un peu mais juste pour tester la connection hardware/software
c'est plus pour avant d'utiliser la voiture une facon de checker tout le hardware (camera, prise de vue, position de la tete, les roues gauche droite etc...)
Ca permet de bien checker que ce qui sera utilise dans le software reste constant par rapport au hardware.
Et pareil le jour de la course en une commande on peut checker ce qu il faut.

File: get_data/get_data_xbox_pad.py (hunk 2 above)
Comment (blopax) [reply get_data/get_data_xbox_pad.py]: juste pour etre sur tu dis ca a cause du get_args() ou simplement le fait de declarer une classe et mettre un if__name__ en dessous ?
sinon yes le but est de faire une pipeline et de se servir de cette pipeline dans un script qui importe les classes. 
J'y penserai en tous cas.

File: get_data/get_data_xbox_pad.py (hunk 1 above)
Comment (blopax) [reply get_data/get_data_xbox_pad.py]: je change ca. Je me l'etais garde pour la fin et j'ai oublie :)

File: process_data/transform_data.py (hunk 3 above)
Comment (blopax) [reply process_data/transform_data.py]: haha t est dur pour l'alphabet. J'avais completement oublié ce detail. mais fair enough. Je change ca. et repush la modif

File: train_data/model_params_setter.py (hunk 5 above)
Comment (blopax) [reply train_data/model_params_setter.py]: ok

**Comment by Talasta** [comment]
> Quelques suggestions en comment. Mais super boulot, le code est clean et agréable a lire :)
> (il manque juste les docstring qui pourrait être utile pour certaines fonctions mais on verra plus tard)

Tu peux valider ? Ou tu veux qu'il corrige ce que tu as noté ?

File: get_data/get_data_xbox_pad.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        for frame in self.camera.capture_continuous(self.rawCapture, format="rgb", use_video_port=True):
+            # convert img as Array
+            image = frame.array
+            # take a pic
+            if self.label[0] != -1:
+                if time.time() - start > self.delay:
+                    im = Image.fromarray(image, 'RGB')
+                    t_stamp = time.time()
+                    picture_path = "/home/pi/Documents/Patate/Pics/Auto/{}_{}_{}.jpg".format(
+                        str(self.label[0]), str(self.label[1]), str(t_stamp))
```
[the remark this answers is not in the document]
Comment (blopax) [reply get_data/get_data_xbox_pad.py]: tout a fait d'accord. Je pense qu on va un peu changer l'archi des docs aussi. On peut en discuter demain voir comment on fait et le changer a ce moment la. J ai mis dans le trello pour etre sur de pas oublier.

File: process_data/transform_data.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    # TODO: For all, make it modular, define a target_directory where you copy renamed data (avoid loosing raw data)
+    # TODO: Use a dictionnary with speed, direction head rather than list that makes code inconsistent
+    #  + ideally just amend the database once images stored in S3
+
+    def five_to_three(self):
+        for filename in os.listdir(self.dir_path):
+            filename_array = filename.split("_")
+            if filename_array[1] == "1":
+                os.rename(os.path.join(self.dir_path, filename),
+                          os.path.join(self.dir_path, "{}_0_{}".format(filename_array[0], filename_array[2])))
```
[the remark this answers is not in the document]
Comment (blopax) [reply process_data/transform_data.py]: nice j avais oublie cette lib :) je l ai mis dans le trello.

File: race.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+    @staticmethod
+    def choose_speed(predictions):
+        if predictions[1] == 1 and predictions[0] == 1:
+            return SPEED_FAST
+        return SPEED_NORMAL
+
+    @staticmethod
+    def choose_head(predictions, speed):
+        if speed == SPEED_FAST and predictions[1] == 1 and predictions[0] == 1:
```
[the remark this answers is not in the document]
Comment (blopax) [reply race.py]: Je ne pense pas car le speed correspond a ce moment a l'image d avant non ? Donc en gros ce qui etait dans le code c'est si 2 predictions de suites disent SPEED_FAST alors HEAD_UP.
apres je pense qu on devra un peu reecrire cette partie mais je voulais pas trop changer la logique existante avant de m'assurer que le code "fonctionne". Et probablement plus commenter la logique dans cette partie...

File: race.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    options = get_args()
+    race_on = RaceOn(options.model_path)
+    print("Ready ! press CTRL+C to START/STOP :")
+
+    try:
+        while True:
+            pass
+    except KeyboardInterrupt:
+        pass
+
```
[the remark this answers is not in the document]
Comment (blopax) [reply race.py]: Je l ai mis dans le trello je pense t as tout a fait raison c bcp plus clean (et en fonction de l'interface http les controls pourront etre affines.

File: train_data/model_params_setter.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        'model_name': "model_race_speed.h5",
+        'model_inputs': [img_in],
+        'model_outputs': [out_speed, out_dir],
+    }
+    return model_parameters
+
+
+def get_images_directory():
+    dataset = "../../../../speed"
+    full_image_directory = dataset + '/Train'
```
[the remark this answers is not in the document]
Comment (blopax) [reply train_data/model_params_setter.py]: idem. j ai mis une carte ds le trello pour ne pas oublier les 2 endroits...

File: race.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jjauzion) [reply race.py]: ok, t'as raison, on fait une 1er version sans modifier la logique et on ajustera après.

Mais vu que `speed` est updaté juste avant l'appel l'appel a la fct `choose_head()`, j'ai bien l'impression que c'est sur l'image courante et non la précédente

[the remark this answers is not in the document]
Comment (blopax) [reply race.py]: c'est toi qui a raison :) bien vu ! je change ca demain. je mettrai le choose head juste avant le speed...
